rockcraft/oci.py

```diff
@@ -320,15 +339,62 @@ def _add_layer_into_image(image_path: Path, archived_content: Path, **kwargs) ->
     _process_run(cmd + ["--history.created_by", " ".join(cmd)])
 
 
-def _archive_layer(layer_path: Path, temp_tar_file: Path) -> None:
+def _archive_layer(
+    layer_path: Path, temp_tar_file: Path, lower_rootfs: Optional[Path] = None
+) -> None:
     """Prepare new OCI layer by archiving its content into tar file.
 
     :param layer_path: path to the content to be archived into a layer
-    :param temp_tar_file: path to the temporary tar fail holding the archived content
+    :param temp_tar_file: path to the temporary tar file holding the archived content
+    :param lower_rootfs: optional path to the filesystem containing the extracted
+        base below this new layer. Used to preserve lower-level directory symlinks,
+        like the ones from Debian/Ubuntu's usrmerge.
     """
     with tarfile.open(temp_tar_file, mode="w") as tar_file:
-        # With `arcname="."` the contents are added relative to `layer_path`.
-        tar_file.add(layer_path, arcname=".")
+        for dirpath, subdirs, filenames in os.walk(layer_path):
+            # Sort `subdirs` in-place, to ensure that `os.walk()` iterates on
+            # them in sorted order.
+            subdirs.sort()
+
+            upper_subpath = Path(dirpath)
+            lower_symlink_target: Optional[Path] = None
+
+            # Handle adding an entry for the directory. We only do that if:
+            # - The directory is not the root (to skip a spurious "." entry);
+            # - The directory's name does not exist on ``lower_rootfs`` as a symlink
+            #   to another directory (like in usrmerge).
+            if upper_subpath != layer_path:
+                add_dir = True
+                relative_path = upper_subpath.relative_to(layer_path)
+
+                if lower_rootfs is not None:
+                    lower_path = lower_rootfs / relative_path
+                    emit.trace(
+                        f"Skipping {upper_subpath} because it exists as a symlink on the lower layer"
+                    )
+                    add_dir = not lower_path.is_symlink()
+                    if lower_path.is_symlink():
+                        lower_symlink_target = Path(os.readlink(lower_path))
+
+                if add_dir:
+                    emit.trace(f"Adding to layer: {upper_subpath}")
+                    tar_file.add(
+                        upper_subpath, arcname=f"./{relative_path}", recursive=False
+                    )
+
+            # Handle adding each file in the directory.
+            for name in filenames:
+                path = upper_subpath / name
+                if lower_symlink_target is not None:
+                    # If the directory containing `name` is a symlink on the lower
+                    # layer, add `name` into the symlink's target instead.
+                    # For example: if the file is `bin/file.txt`, and `bin` is a
+                    # symlink to `usr/bin`, add the file as `usr/bin/file.txt`.
```

Comment:
If we implement a separate utility function to resolve the symlink based on the base filesystem we could later improve it to resolve symlinks to symlinks without changing the caller code.